File: ruby/lib/steam/steam_player.rb
# This code is free software; you can redistribute it and/or modify it under
# the terms of the new BSD License.
#
# Copyright (c) 2008-2011, Sebastian Staudt

require 'exceptions/steam_condenser_exception'

# The SteamPlayer class represents a player connected to a server
#
# @author Sebastian Staudt
class SteamPlayer

  # Returns the client port of this player
  #
  # @return [Fixnum] The client port of the player
  attr_reader :client_port

  # Returns the time this player is connected to the server
  #
  # @return [Float] The connection time of the player
  attr_reader :connect_time

  # Returns the ID of this player
  #
  # @return [Fixnum] The ID of this player
  attr_reader :id

  # Returns the IP address of this player
  #
  # @return [String] The IP address of this player
  attr_reader :ip_address

  # Returns the nickname of this player
  #
  # @return [String] The name of this player
  attr_reader :name

  # Returns the packet loss of this player's connection
  #
  # @return [String] The packet loss of this player's connection
  attr_reader :loss

  # Returns the ping of this player
  #
  # @return [Fixnum] The ping of this player
  attr_reader :ping

  # Returns the rate of this player
  #
  # @return [Fixnum] The rate of this player
  attr_reader :rate

  # Returns the real ID (as used on the server) of this player
  #
  # @return [Fixnum] The real ID of this player
  attr_reader :real_id

  # Returns the score of this player
  #
  # @return [Fixnum] The score of this player
  attr_reader :score

  # Returns the connection state of this player
  #
  # @return [String] The connection state of this player
  attr_reader :state

  # Returns the SteamID of this player
  #
  # @return [String] The SteamID of this player
  attr_reader :steam_id

  # Creates a new player instance with the given information
  #
  # @param [Fixnum] id The ID of the player on the server
  # @param [String] name The name of the player
  # @param [Fixnum] score The score of the player
  # @param [Float] connect_time The time the player is connected to the server
  def initialize(id, name, score, connect_time)
    @connect_time = connect_time
    @id = id
    @name = name
    @score = score
    @extended = false
  end

  # Extends a player object with information retrieved from a RCON call to
  # the status command
  #
  # @param [String] player_data The player data retrieved from `rcon status`
  # @raise [SteamCondenserException] if the information belongs to another
  #         player
  def add_info(player_data)
    unless player_data[:name] == @name
      raise SteamCondenserException, 'Information to add belongs to a different player.'
    end

    @extended = true

    @real_id  = player_data[:userid].to_i
    @steam_id = player_data[:uniqueid]
    @state    = player_data[:state] if player_data.key? :state

    if !bot?
      @loss = player_data[:loss].to_i
      @ping = player_data[:ping].to_i

      if player_data.key? :adr
        @ip_address, @client_port  = player_data[:adr].split(':')
        @client_port = @client_port.to_i
      end

      @rate  = player_data[:rate].to_i if player_data.key? :rate
    end
  end

  # Returns whether this player is a bot
  #
  # @return bool `true` if this player is a bot
  def bot?
    @steam_id == 'BOT'
  end

  # Returns whether this player object has extended information gathered
  # using RCON
  #
  # @return bool `true` if extended information for this player is available
  def extended?
    @extended
  end

  # Returns a string representation of this player
  #
  # @return [String] A string representing this player
  def to_s
    if @extended
      "\##{@real_id} \"#{@name}\", SteamID: #{@steam_id}, Score: #{@score}, Time: #{@connect_time}"
    else
      "\##{@id} \"#{@name}\", Score: #{@score}, Time: #{@connect_time}"
    end
  end

end

File: php/lib/steam/SteamPlayer.php
<?php
/**
 * This code is free software; you can redistribute it and/or modify it under
 * the terms of the new BSD License.
 *
 * Copyright (c) 2008-2011, Sebastian Staudt
 *
 * @author     Sebastian Staudt
 * @license    http://www.opensource.org/licenses/bsd-license.php New BSD License
 * @package    Steam Condenser (PHP)
 * @subpackage SteamPlayer
 */

require_once STEAM_CONDENSER_PATH . 'exceptions/SteamCondenserException.php';

/**
 * @package Steam Condenser (PHP)
 * @subpackage SteamPlayer
 */
class SteamPlayer
{
    /**
     * @var int
     */
    private $clientPort;

    /**
     * @var float
     */
    private $connectTime;

    /**
     * @var boolean
     */
    private $extended;

    /**
     * @var int
     */
    private $id;

    /**
     * @var String
     */
    private $ipAddress;

    /**
     * @var int
     */
    private $loss;

   /**
    * @var String
    */
    private $name;

    /**
     * @var int
     */
    private $ping;

    /**
     * @var int
     */
    private $realId;

    /**
     * @var int
     */
    private $score;

    /**
     * @var String
     */
    private $state;

    /**
     * @var String
     */
    private $steamId;

    /**
     * Creates a new SteamPlayer object with the given information
     * @param int $playerId
     * @param String $playerName
     * @param int $playerPoints
     * @param float $playerConnectTime
     */
    public function __construct($id, $name, $score, $connectTime)
    {
        if(!is_int($id) || $id < 0)
        {
            throw new Exception("Player ID has to be a number greater than or equal 0.");
        }

        if(!is_string($name))
        {
            throw new Exception("Player name has to be a string.");
        }

        if(!is_int($score))
        {
            throw new Exception("Player points have to be a number.");
        }

        if(!is_float($connectTime))
        {
            throw new Exception("Player connection time has to be a floating-point integer.");
        }

        $this->connectTime = $connectTime;
        $this->id = $id;
        $this->name = $name;
        $this->score = $score;
        $this->extended = false;
    }

    /**
     * Extends a player object with information retrieved from a RCON call to
     * the status command
     *
     * @param string playerData The player data retrieved from
     *        <code>rcon status</code>
     * @throws SteamCondenserException
     */
    public function addInformation($playerData) {
        if($playerData['name'] != $this->name) {
            throw new SteamCondenserException('Information to add belongs to a different player.');
        }

        $this->extended = true;
        $this->realId   = intval($playerData['userid']);
        if(array_key_exists('state', $playerData)) {
            $this->state    = $playerData['state'];
        }
        $this->steamId  = $playerData['uniqueid'];

        if(!$this->isBot()) {
            $this->loss = intval($playerData['loss']);
            $this->ping = intval($playerData['ping']);

            if(array_key_exists('data', $playerData)) {
                $address = explode(':', $playerData['adr']);
                $this->ipAddress  = $address[0];
                $this->clientPort = intval($address[1]);
            }

            if(array_key_exists('rate', $playerData)) {
                $this->rate = $playerData['rate'];
            }
        }
    }

    /**
     * Returns the client port of this player
     * @return int
     */
    public function getClientPort()
    {
        return $this->clientPort;
    }

    /**
     * Returns the time this player is connected to the server
     * @return float
     */
    public function getConnectTime()
    {
        return $this->connectTime;
    }

    /**
     * Returns the ID of this player
     * @return int
     */
    public function getId()
    {
        return $this->id;
    }

    /**
     * Returns the IP address of this player
     * @return String
     */
    public function getIpAddress()
    {
        return $this->ipAddress;
    }

    /**
     * Returns the nickname of this player
     * @return String
     */
    public function getName()
    {
        return $this->name;
    }

    /**
     * Returns the ping of this player
     * @return int
     */
    public function getPing()
    {
        return $this->ping;
    }

    /**
     * @return Returns the rate of this player
     */
    public function getRate() {
        return $this->rate;
    }

    /**
     * Returns the real ID (as used on the server) of this player
     * @return int
     */
    public function getRealId()
    {
        return $this->realId;
    }

    /**
     * Returns the score of this player
     * @return int
     */
    public function getScore()
    {
        return $this->score;
    }

    /**
     * Returns the connection state of this player
     * @return String
     */
    public function getState()
    {
        return $this->state;
    }

    /**
     * Returns the SteamID of this player
     * @return String
     */
    public function getSteamId()
    {
        return $this->steamId;
    }

    /**
     * Returns whether this player object has extended information
     * @return boolean
     */
    public function isBot() {
        return $this->steamId == 'BOT';
    }

    /**
     * Returns whether this player object has extended information gathered
     * using RCON
     * @return boolean
     */
    public function isExtended()
    {
        return $this->extended;
    }

    /**
     * Returns a String representation of this player
     * @return String
     */
    public function __toString()
    {
        if($this->extended) {
            return "#{$this->realId} \"{$this->name}\", SteamID: {$this->steamId} Score: {$this->score}, Time: {$this->connectTime}";
        }
        else {
            return "#{$this->id} \"{$this->name}\", Score: {$this->score}, Time: {$this->connectTime}";
        }
    }
}
?>
